google/cloud/internal/url_encode.cc

```diff
@@ -24,15 +24,31 @@ namespace internal {
 std::string UrlEncode(absl::string_view value) {
   using CharMapping = std::pair<absl::string_view, absl::string_view>;
   auto const mappings = std::array<CharMapping, 25>{{
-      {" ", "%20"}, {"\"", "%22"}, {"#", "%23"},  {"$", "%24"}, {"%", "%25"},
-      {"&", "%26"}, {"+", "%2B"},  {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},
-      {";", "%3B"}, {"<", "%3C"},  {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},
-      {"@", "%40"}, {"[", "%5B"},  {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
-      {"`", "%60"}, {"{", "%7B"},  {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"},
+      {" ", "%20"}, {"\"", "%22"}, {"#", "%23"}, {"$", "%24"},    {"&", "%26"},
+      {"+", "%2B"}, {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},    {";", "%3B"},
+      {"<", "%3C"}, {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},    {"@", "%40"},
+      {"[", "%5B"}, {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},    {"`", "%60"},
+      {"{", "%7B"}, {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"}, {"%", "%25"},
   }};
   return absl::StrReplaceAll(value, mappings);
 }
 
+std::string UrlDecode(absl::string_view value) {
+  using CharMapping = std::pair<absl::string_view, absl::string_view>;
+  // Note: absl::StrReplaceAll applies mappings greedily, therefore the %
+  // mapping must be placed last for decoding.
+  auto mappings = std::array<CharMapping, 25>{{
+      {" ", "%20"}, {"\"", "%22"}, {"#", "%23"}, {"$", "%24"},    {"&", "%26"},
+      {"+", "%2B"}, {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},    {";", "%3B"},
+      {"<", "%3C"}, {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},    {"@", "%40"},
+      {"[", "%5B"}, {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},    {"`", "%60"},
+      {"{", "%7B"}, {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"}, {"%", "%25"},
+  }};
+  std::for_each(mappings.begin(), mappings.end(),
+                [&](CharMapping pair) { std::swap(pair.first, pair.second); });
```

Comment:
I worked up a version of `UrlDecode()` yesterday, and concluded that (until we have a C++17 minimum and can constexpr build one mapping from the other) we are best off with two arrays and no run-time overhead.  We can discuss offline if you like.

I also added some test cases.